internal/controller/engine/engine.go

```diff
@@ -0,0 +1,411 @@
+		<-e.mgr.Elected()
+
+		e.log.Debug("Starting new controller", "controller", name)
+
+		// Run the controller until its context is cancelled.
+		if err := c.Start(ctx); err != nil {
+			e.log.Info("Controller returned an error", "name", name, "error", err)
+		}
+
+		e.log.Debug("Stopped controller", "controller", name)
+	}()
+
+	if co.gc != nil {
+		go func() {
+			// Don't start the garbage collector until the manager is elected.
+			<-e.mgr.Elected()
+
+			e.log.Debug("Starting watch garbage collector for controller", "controller", name)
+
+			// Run the collector every minute until its context is cancelled.
+			co.gc.GarbageCollectWatches(ctx, 1*time.Minute)
+
+			e.log.Debug("Stopped watch garbage collector for controller", "controller", name)
+		}()
+	}
+
+	r := &controller{
+		ctrl:    c,
+		cancel:  cancel,
+		sources: make(map[schema.GroupVersionKind]*StoppableSource),
+	}
+
+	e.mx.Lock()
+	e.controllers[name] = r
+	e.mx.Unlock()
+
+	return nil
+}
+
+// Stop a controller.
+func (e *ControllerEngine) Stop(ctx context.Context, name string) error {
+	e.mx.RLock()
+	c, running := e.controllers[name]
+	e.mx.RUnlock()
+
+	// Stop is a no-op if the controller isn't running.
+	if !running {
+		return nil
+	}
+
+	// First stop the controller's watches.
+	if err := e.StopWatches(ctx, name); err != nil {
+		return errors.Wrapf(err, "cannot stop watches for controller %q", name)
+	}
+
+	// Then stop any informers that only this controller was using.
+	if err := e.RemoveUnwatchedInformers(ctx); err != nil {
+		return errors.Wrapf(err, "cannot remove unwatched informers after stopping watches for controller %q", name)
+	}
+
+	// Finally, stop and delete the controller.
+	e.mx.Lock()
+	c.cancel()
+	delete(e.controllers, name)
+	e.mx.Unlock()
+
+	e.log.Debug("Stopped controller", "controller", name)
+	return nil
+}
+
+// IsRunning returns true if the named controller is running.
+func (e *ControllerEngine) IsRunning(name string) bool {
+	e.mx.RLock()
+	_, running := e.controllers[name]
+	e.mx.RUnlock()
+	return running
+}
+
+// Watch an object.
+type Watch struct {
+	kind       client.Object
+	handler    handler.EventHandler
+	predicates []predicate.Predicate
+}
+
+// WatchFor returns a Watch for the supplied kind of object. Events will be
+// handled by the supplied EventHandler, and may be filtered by the supplied
+// predicates.
+func WatchFor(kind client.Object, h handler.EventHandler, p ...predicate.Predicate) Watch {
+	return Watch{kind: kind, handler: h, predicates: p}
+}
+
+// StartWatches instructs the named controller to start the supplied watches.
+// The controller will only start a watch if it's not already watching the type
+// of object specified by the supplied Watch. StartWatches blocks other
+// operations on the same controller if and when it starts a watch.
+func (e *ControllerEngine) StartWatches(name string, ws ...Watch) error {
+	e.mx.RLock()
+	c, ok := e.controllers[name]
+	e.mx.RUnlock()
```

Comment:
Note there's two mutexes. One on the engine, one on each controller.

Here we:

* Briefly take a read lock on the whole engine to get the named controller
* Take a read lock on the controller to figure out what watches (if any) we need to start.
* Take a write lock on the controller only if there are watches to start.

This means when an XR reconciler calls `StartWatches` it'll only block other XRs powered by the same controller - i.e. XRs of the same type, not XRs of any type. It'll only block them if it's composing a kind of resource that no other XR of its type has composed before.